ruby/ql/lib/codeql/ruby/ApiGraphs.qll

```diff
@@ -257,6 +258,29 @@ module API {
      */
     Node getAnImmediateSubclass() { result = this.getASuccessor(Label::subclass()) }
 
+    /**
+     * Gets a node representing the `content` stored on the base object.
+     */
+    Node getContent(DataFlow::Content content) {
+      result = this.getASuccessor(Label::content(content))
```

Comment:
It is done like this exactly to avoid the fan-out from `getAReadContent`.

Here's a break down of how the graph is generated and what ends up being matched:
| Description | Edge label  | Matched by  |
|---|---|---|
| Use node reading a known element `n`   | `TKnownElementContent(n)`  | `Element[n]` and `Element[any]`  |
| Def node storing a known element `n`   | `TKnownElementContent(n)`  | `Element[n]` and `Element[any]`  |
| Use node reading an unknown element   | `TUnknownElementContent`  | `Element[?]` and `Element[any]`  |
| Def node storing an unknown element   | `TUnknownElementContent`  | `Element[?]` and `Element[any]`  |

`Element[n]` does not match `TUnknownElementContent`, and the model must thus use `Element[?,n]` if this behaviour is wanted. This seems consistent with how existing Ruby models are required to mention `Element[?]` explicitly.

I believe this is the best integration we can hope for without changing how `ContentSet` works.

I tried storing `ContentSet` on the edge label but it doesn't really work out nicely. `getAReadContent()` returns a `Content`, but the edges are labelled with a `ContentSet` so we have to map it back to a `ContentSet` to look up in the `edge` relation. To ensure an efficient join there we'd end up factoring out a relation indexed on the `Content`, which basically corresponds the edge relation we're using now.

Reply:
Thanks for the detailed explanation.

Reply:
After merging with https://github.com/github/codeql/pull/10574 I believe the table is updated as follows:
- the first two rows can also be matched by `Element[n!]`
- the last two rows can also be matched by `Element[k]` for any constant `k`

| Description | Edge label  | Matched by  |
|---|---|---|
| Use node reading a known element `n`   | `TKnownElementContent(n)`  | `Element[n]`, `Element[n!]`, and `Element[any]`  |
| Def node storing a known element `n`   | `TKnownElementContent(n)`  | `Element[n]`, `Element[n!]`, and `Element[any]`  |
| Use node reading an unknown element   | `TUnknownElementContent`  | `Element[k]` for any `k`, `Element[?]` and `Element[any]`  |
| Def node storing an unknown element   | `TUnknownElementContent`  | `Element[k]` for any `k`, `Element[?]` and `Element[any]`  |

- Previously it was the model's responsibility to mention `Element[?]` if unknown reads/stores should be found.
- Now models are now responsible for using `Element[n!]` if this behaviour is _not_ wanted.

Again this seems consistent with how Ruby models work after https://github.com/github/codeql/pull/10574, so I'm cautiously optimistic that everything just works after this change.